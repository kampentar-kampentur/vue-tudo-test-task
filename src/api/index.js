import axios from 'axios'

// keep application base URL blank for proper work of reverse-proxy (http-proxy-middleware) in vue.config.js
// needed for local development because of CORS restrictions
export const jsonPlaceholderAPI = axios.create({
  baseURL: process.env.VUE_APP_BASE_API_URL
})

jsonPlaceholderAPI.interceptors.response.use(
  response => response,
  error => {
    console.error(error)
    return Promise.reject(error)
  }
)

export default axios
