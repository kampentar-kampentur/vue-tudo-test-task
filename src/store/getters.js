export default {
  'getAllTodoList': state =>state.todoList
}
