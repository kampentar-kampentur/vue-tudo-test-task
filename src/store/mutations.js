export default {
  'updateTodoList'(state, payload) {
    Object.assign(state, { todoList: payload })
  },
  'updateTodoItem'(state, payload) {
    Object.assign(state.todoList[payload.index], payload.todoItem)
  }
}
