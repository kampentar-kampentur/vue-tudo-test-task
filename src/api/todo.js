import { jsonPlaceholderAPI } from '../api'

export const getAllTodoList = userId => jsonPlaceholderAPI.get(`/users/${userId}/todos`)
export const addTodoItemToListItem = todoItem => jsonPlaceholderAPI.post(`/todos`, todoItem)
export const deleteTodoFromList = todoId => jsonPlaceholderAPI.delete(`/todos/${todoId}`)
export const changeTodoItem = todoItem => jsonPlaceholderAPI.put('/todos/1', todoItem)
