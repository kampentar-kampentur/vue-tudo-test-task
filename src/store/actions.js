/* eslint-disable */
import { getAllTodoList, addTodoItemToListItem, deleteTodoFromList, changeTodoItem } from '@/api/todo'

export default {
  async 'restoreTodoList' (context, payload) {
    let response = await getAllTodoList (payload)
    context.commit('updateTodoList', response.data)
  },
  async 'addTodoItemToList' (context, payload) {
    await addTodoItemToListItem(payload)
    const todoList = [...context.state.todoList, payload]
    context.commit('updateTodoList', todoList)
  },
  async 'removeTodoItemFromList' (context, payload) {
    const todoList = [...context.state.todoList ]
    const index = todoList.findIndex(el => el.id === payload)
    if (index === -1) return
    await deleteTodoFromList(payload)
    todoList.splice(index, 1)
    context.commit('updateTodoList', todoList)
  },
  async 'changeTodoItemState' (context, payload) {
    const todoList = [ ...context.state.todoList ]
    const index = todoList.findIndex(el => el.id === payload.id)
    if (index === -1) return
    await changeTodoItem(payload)
    context.commit('updateTodoItem', { index, todoItem: payload})
  }
}
