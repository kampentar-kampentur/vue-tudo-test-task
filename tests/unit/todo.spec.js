import axios from 'axios';
import Vuex from 'vuex';
import ElementUI from 'element-ui';
import flushPromises from 'flush-promises';
import registerDirectives from '../../src/directives'
import { axiosMock } from '../mocks/api';
import { mount, createLocalVue  } from '@vue/test-utils';

import { mockToDoApi } from '../mocks/api/todo.api.mock';

import store from '../../src/store'
import TodoPageComponent from '../../src/views/todoPage/TodoPage.vue';
import * as pageObject from './todo.po';

const localVue = createLocalVue()

localVue.use(Vuex);
localVue.use(ElementUI);
registerDirectives(localVue)

let toDoWrapper;

beforeEach(() => {
  mockToDoApi();
});

beforeEach(() => {
  toDoWrapper = mount(TodoPageComponent, { store, localVue });
});

afterAll(() => {
  axiosMock.reset();
});

describe('api', () => {
  it('should read items', async () => {
    const itemsResponse = await axios.get('/users/1/todos');
    expect(itemsResponse.data).toEqual([
        {
          "userId": 1,
          "id": 1,
          "title": "delectus aut autem",
          "completed": false
        },
        {
          "userId": 1,
          "id": 2,
          "title": "quis ut nam facilis et officia qui",
          "completed": false
        }
    ]);
  });

  it('should add items', async () => {
    const newItem = {
      "userId": 1,
      "id": 3,
      "title": "My new item",
      "completed": true
    };
    const itemsResponse = await axios.post('/todos', newItem);
    expect(itemsResponse.data).toEqual(newItem);
  });

  it('should remove items', async () => {
    const itemsResponse = await axios.delete('/todos/2');
    expect(itemsResponse.data).toEqual({
      "userId": 1,
      "id": 2,
      "title": "quis ut nam facilis et officia qui",
      "completed": false
    });
  });

  it('should update item', async () => {
    const itemsResponse = await axios.put('/todos/2', {
      "title": "Edited Item",
      "completed": true
    });
    expect(itemsResponse.data).toEqual({
      "userId": 1,
      "id": 2,
      "title": "Edited Item",
      "completed": true
    });
  });
});

describe('Todo component', () => {
  it('should pass sanity check', () => {
    expect(toDoWrapper.attributes().class).toBe('todo-page');
  });
  
  it('should match snapshot', () => {
    expect(toDoWrapper.element).toMatchSnapshot()
  });

  it('should add new item', async () => {
    await flushPromises();
    const itemEls = pageObject.getItems(toDoWrapper);
    const newInputEl = pageObject.newItemInput(toDoWrapper);
    const addButton = pageObject.addButton(toDoWrapper);
    expect(itemEls.length).toBe(2);
    newInputEl.setValue('My item from test');
    await localVue.nextTick();
    addButton.trigger('click');
    await flushPromises();
    expect(pageObject.getItems(toDoWrapper).length).toBe(3);
  });
  it('should remove item', async () => {
    await flushPromises();
    const itemEls = pageObject.getItems(toDoWrapper);
    const itemEl = pageObject.getItem(toDoWrapper, 1);
    const deleteButton = pageObject.getDeleteBtn(itemEl);
    expect(itemEls.length).toBe(2);
    deleteButton.trigger('click');
    await flushPromises();
    expect(pageObject.getItems(toDoWrapper).length).toBe(1);
  });
  it('should update checked state', async () => {
    await flushPromises();
    const itemEls = pageObject.getItems(toDoWrapper);
    const itemEl = pageObject.getItem(toDoWrapper, 1);
    const stateCheckboxLabel = pageObject.getStateCheckboxLabel(itemEl);
    const stateCheckbox = pageObject.getStateCheckbox(stateCheckboxLabel);
    expect(itemEls.length).toBe(2);
    expect(stateCheckbox.element.checked).toBe(false);
    stateCheckboxLabel.trigger('click');
    await flushPromises();
    expect(pageObject.getItems(toDoWrapper).length).toBe(2);
    expect(pageObject.getStateCheckbox(itemEl).element.checked).toBe(true);
  });
  it('should update title', async () => {
    await flushPromises();
    let itemEl = pageObject.getItem(toDoWrapper, 1);
    let itemLabel = pageObject.getItemLabel(itemEl);
    const updateTitleBtn = pageObject.getUpdateTitleBtn(itemEl);
    expect(itemLabel.text()).toBe('quis ut nam facilis et officia qui');
    updateTitleBtn.trigger('click');
    await localVue.nextTick();
    const updateTitleInput = pageObject.getUpdateTitleInput(itemEl);
    updateTitleInput.setValue('Updated item');
    updateTitleInput.trigger('keypress', {key: 'Enter'});
    await flushPromises();
    itemEl = pageObject.getItem(toDoWrapper, 1)
    itemLabel = pageObject.getItemLabel(itemEl);
    expect(itemLabel.text()).toBe('Updated item');
  });
});