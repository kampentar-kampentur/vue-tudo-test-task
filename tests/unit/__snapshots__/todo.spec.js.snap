// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo component should match snapshot 1`] = `
<div
  class="todo-page"
>
  <div
    class="container"
  >
    <div
      class="add-todo-wrapper"
    >
      <div
        class="el-input"
      >
        <!---->
        <input
          autocomplete="off"
          class="el-input__inner"
          id="new-item-input"
          placeholder="Please input"
          type="text"
        />
        <!---->
        <!---->
        <!---->
        <!---->
      </div>
       
      <button
        class="el-button el-button--primary is-disabled"
        disabled="disabled"
        id="add-new-button"
        type="button"
      >
        <!---->
        <!---->
        <span>
          Add
        </span>
      </button>
    </div>
     
    <div
      class="todo-items-wrapper"
    >
      <div
        class="todo-item"
      >
        <label
          class="el-checkbox todo-item-checkbox"
        >
          <span
            class="el-checkbox__input"
          >
            <span
              class="el-checkbox__inner"
            />
            <input
              aria-hidden="false"
              class="el-checkbox__original"
              type="checkbox"
              value=""
            />
          </span>
          <span
            class="el-checkbox__label"
          >
            <span
              class="todo-item-label"
            >
              delectus aut autem
            </span>
            <!---->
          </span>
        </label>
         
        <div
          class="todo-item-buttons-wrapper"
        >
          <button
            class="el-button todo-item-edit-btn el-button--primary"
            type="button"
          >
            <!---->
            <i
              class="el-icon-edit"
            />
            <!---->
          </button>
           
          <button
            class="el-button todo-item-remove-btn el-button--warning"
            type="button"
          >
            <!---->
            <i
              class="el-icon-delete"
            />
            <!---->
          </button>
        </div>
      </div>
      <div
        class="todo-item"
      >
        <label
          class="el-checkbox todo-item-checkbox"
        >
          <span
            class="el-checkbox__input"
          >
            <span
              class="el-checkbox__inner"
            />
            <input
              aria-hidden="false"
              class="el-checkbox__original"
              type="checkbox"
              value=""
            />
          </span>
          <span
            class="el-checkbox__label"
          >
            <span
              class="todo-item-label"
            >
              quis ut nam facilis et officia qui
            </span>
            <!---->
          </span>
        </label>
         
        <div
          class="todo-item-buttons-wrapper"
        >
          <button
            class="el-button todo-item-edit-btn el-button--primary"
            type="button"
          >
            <!---->
            <i
              class="el-icon-edit"
            />
            <!---->
          </button>
           
          <button
            class="el-button todo-item-remove-btn el-button--warning"
            type="button"
          >
            <!---->
            <i
              class="el-icon-delete"
            />
            <!---->
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
